packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/cloudwatch-logs.ts

```diff
@@ -0,0 +1,102 @@
+import * as ecs from '@aws-cdk/aws-ecs';
+import * as awslogs from '@aws-cdk/aws-logs';
+import * as cdk from '@aws-cdk/core';
+import { Service } from '../service';
+import { Container } from './container';
+import { ContainerMutatingHook, ServiceExtension } from './extension-interfaces';
+
+/**
+ * Settings for the hook which mutates the application container
+ * to send logs to CloudWatch Logs
+ */
+export interface CloudWatchLogsProps {
+  /**
+   * The parent service that is being mutated
+   */
+  readonly parentService: Service;
+
+  /**
+   * The log group into which logs should be routed
+   */
+  readonly logGroup: awslogs.LogGroup;
+}
+
+/**
+ * This hook modifies the application container's settings so that
+ * it routes logs using FireLens
+ */
+export class CloudWatchLogsMutatingHook extends ContainerMutatingHook {
+  private parentService: Service;
+  private logGroup: awslogs.LogGroup;
+
+  constructor(props: CloudWatchLogsProps) {
+    super();
+    this.parentService = props.parentService;
+    this.logGroup = props.logGroup;
+  }
+
+  public mutateContainerDefinition(props: ecs.ContainerDefinitionOptions) {
+    return {
+      ...props,
+      logging: new ecs.AwsLogDriver({
+        streamPrefix: this.parentService.id,
+        logGroup: this.logGroup,
+      }),
+    } as ecs.ContainerDefinitionOptions;
+  }
+}
+
+/**
+ * Settings for the hook which mutates the application container
+ * to send logs to CloudWatch Logs
+ */
+export interface CloudWatchLogsExtensionProps {
+  /**
+   * The log group into which logs should be routed
+   */
+  readonly logGroup?: awslogs.LogGroup;
+}
+
+/**
+ * This extension modifies a service to send its application logs to CloudWatch
+ * logs
+ */
+export class CloudWatchLogsExtension extends ServiceExtension {
+  private logGroup!: awslogs.LogGroup;
```

Comment:
Not sure if it makes sense to put `!` (non-nullable) here. Technically this property can be null if the constructor is the only method invoked (if the service implementation changes in the future).

if the intent is to make this non-nullable, then we should be adding the logic from the prehook into the constructor.

Reply:
Unfortunately one downside of using the API like:

`service.add(new MyExtension())`

is that the extension does not know what service it will be added to at the time of constructor. It only knows the service later on. This is why some things don't happen until the prehook is called by the service. In this case I want the log group to have the name of the service in it so we wait until the prehook to create the log group.

More generally I want to avoid putting too much logic into the constructor because the constructor executes at the time of instantiating that one extension, vs the prehook executes after all the extensions have been added, so the extension can make better decisions in the prehook as it has the entire state built up at that time.

I have not found a good solution to this problem of properties that will be used later in the prehook, but start out undefined at constructor time. Any suggestion you might have to make this more clear I'd love to consider though

Reply:
Makes sense, I'd probably suggest removing the ! since there's no guarantee that this object is not null, and forcing any callers down the line to do a null check.